src/Couchbase.Lite.Shared/API/Database/Collection.cs

```diff
@@ -269,13 +321,102 @@ public void DeleteIndex([NotNull] string name)
 
         #endregion
 
-        #region IDispose
+        #region Public Methods - QueryFactory
 
-        public void Dispose()
+        public IQuery CreateQuery([NotNull] string queryExpression)
         {
             throw new NotImplementedException();
         }
 
         #endregion
+
+        #region Internal Methods
+
+        /// <summary>
+        /// Returns false if this collection has been deleted, or its database closed.
+        /// </summary>
+        internal void CheckCollectionValid()
+        {
+            ThreadSafety.DoLocked(() =>
+            {
+                if (c4Db == null) {
+                    throw new InvalidOperationException(CouchbaseLiteErrorMessage.DBClosed);
+                }
+
+                if (_c4coll == IntPtr.Zero) {
+                    throw new InvalidOperationException(String.Format(CouchbaseLiteErrorMessage.CollectionNotAvailable,
+                                ToString()));
+                }
+
+                if (!Native.c4coll_isValid((C4Collection*)_c4coll)) {
+                    throw new InvalidOperationException(String.Format(CouchbaseLiteErrorMessage.CollectionNotAvailable,
+                                ToString()));
+                }
+            });
+        }
+
+        #endregion
+
+        #region Private Methods
+
+        private C4Database* GetC4Database()
+        {
+            C4Database* c4db = null;
+            ThreadSafety.DoLocked(() =>
+            {
+                if (c4coll == null)
+                    return;
+
+                c4db = Native.c4coll_getDatabase(c4coll);
+            });
+
+            return c4db;
+        }
+
+        private unsafe void ReleaseCollection()
+        {
+            Interlocked.Exchange(ref _c4coll, IntPtr.Zero);
+            Native.c4coll_release((C4Collection*)_c4coll);
```

Comment:
This is incorrect, `Interlocked.Exchange` will return the value that was *previously* in the ref variable.  As it stands now, this is a leak since you set _c4coll to 0, and then free _c4coll (which is zero...).  The logic should look like this:

```
var old = Interlocked.Exchange(ref _c4coll, IntPtr.Zero);
Native.c4coll_release((C4Collection *)old);
```